app/Actions/AlbumAuthorisationProvider.php

```
@@ -0,0 +1,523 @@
+			};
+
+			if ($model instanceof BaseAlbumImpl) {
+				// If the queried model is the base class, we can directly
+				// apply the sub-query
+				return $query->where($visibilitySubQuery);
+			} else {
+				// If the queried model is not the base class, but a derived one,
+				// we must apply the sub-query to the relation.
+				return $query->whereHas('base_class', $visibilitySubQuery);
+			}
+		}
+
+		$userID = AccessControl::id();
+
+		// We must wrap everything into an outer query to avoid any undesired
+		// effects in case that the original query already contains an
+		// "OR"-clause.
+		// The sub-query only uses properties (i.e. columns) which are
+		// defined on the common base model for all albums.
+		$visibilitySubQuery = function (Builder $query2) use ($userID) {
+			$query2
+				->where('owner_id', '=', $userID)
+				->orWhere(fn (Builder $q) => $q
+					->where('requires_link', '=', false)
+					->whereHas(
+						'shared_with',
+						fn (Builder $q2) => $q2->where('user_id', '=', $userID)
+					)
+				)
+				->orWhere(fn (Builder $q) => $q
+					->where('requires_link', '=', false)
+					->where('is_public', '=', true)
+				);
+		};
+
+		if ($model instanceof BaseAlbumImpl) {
+			// If the queried model is the base class, we can directly
+			// apply the sub-query
+			return $query->where($visibilitySubQuery);
+		} else {
+			// If the queried model is not the base class, but a derived one,
+			// we must apply the sub-query to the relation.
+			return $query->whereHas('base_class', $visibilitySubQuery);
+		}
+	}
+
+	/**
+	 * Checks whether the album with the given ID is visible for the current
+	 * user.
+	 *
+	 * For real albums (i.e. albums that are stored in the DB), see
+	 * {@link AlbumAuthorisationProvider::applyVisibilityFilter()} for a
+	 * specification of the rules when an album is visible.
+	 * In other cases, the following holds:
+	 *  - the root album is visible if and only if the user is authenticated
```

I don't understand. Shouldn't the root album be visible for _everybody_? Sure, its content will vary from user to user, and if there are no visible top-level albums, it may be empty for unauthenticated users, but I would think it must still be visible?

---

Good point. Actually, the case that something queries for the visibility of the root album never occurs. As by my comment "visibility of an album" means "an album is called _visible_ if the current user is allowed to see the album (itself) within a listing or similar". However, the root album will never be part of listing, because it is the top most album. So it will never be listed as a child album of another album.

Hence, the method handles a case which should never happen. Maybe it would be better to explicitly exclude this case and throw an `InvalidDomainException` if the method is called for the root album.

---

Yeah, I think it's confusing right now. Not only does it handle a case that never happens, it seems to handle it _badly_. That's super-confusing. I think throwing an exception would be a lot better...

---

The whole method `isVisible` is gone now. It has never been called and thus actually has been dead code.